src/Hamiltonians/abstract.jl

```diff
@@ -0,0 +1,257 @@
+```
+"""
+function dimension(::Type{T}, ::BoseFS{N,M}) where {N,M,T<:Integer}
+    try
+        return T(binomial(T(N + M - 1), T(N)))
+    catch
+        return nothing
+    end
+end
+function dimension(::Type{T}, ::BoseFS{N,M}) where {N,M,T<:AbstractFloat}
+    n = N + M - 1
+    k = N
+    try
+        T(binomial(Int128(n), Int128(k)))
+    catch
+        T(exp(logbinomialapprox(n, k)))
+    end
+end
+function dimension(::Type{T}, b::BoseFS2C) where T
+    return dimension(T, b.bsa) * dimension(T, b.bsb)
+end
+
+dimension(h::AbstractHamiltonian) = dimension(Int, h)
+dimension(::Type{T}, h::AbstractHamiltonian) where {T} = dimension(T, starting_address(h))
+
+BitStringAddresses.nearUniform(h::AbstractHamiltonian) = nearUniform(typeof(starting_address(h)))
+
+"""
+    starting_address(h)
+
+Return a starting address for Hamiltonian `h`.
+
+# Example
+
+```jldoctest
+julia> addr = BoseFS((3, 2, 1));
+
+julia> H = HubbardMom1D(addr);
+
+julia> addr == starting_address(H)
+true
+```
+"""
+starting_address
+
+"""
+    Hamiltonians.LOStructure(op::AbstractHamiltonian)
+    Hamiltonians.LOStructure(typeof(op))
+
+`LOStructure` speficies properties of the linear operator `op`. If a special structure is
+known this can speed up calculations. Implemented structures are:
+
+* `Hamiltonians.HermitianLO` The operator is complex and hermitian or real and symmetric.
+* `Hamiltonians.ComplexLO` The operator has no known specific structure.
+
+In order to define this trait for a new linear operator type, define a method for
+`LOStructure(::Type{<:MyNewLOType}) = …`.
+
+"""
+abstract type LOStructure end
+
+struct HermitianLO <: LOStructure end
+struct ComplexLO <: LOStructure end
+
+# defaults
+LOStructure(op::AbstractHamiltonian) = LOStructure(typeof(op))
+LOStructure(::Type{<:AbstractHamiltonian}) = ComplexLO()
+
+"""
+    rayleigh_quotient(H, v)
+
+```math
+\\frac{⟨ v | H | v ⟩}{⟨ v|v ⟩}
+```
+"""
+rayleigh_quotient(lo, v) = dot(v, lo, v)/norm(v)^2
+
+"""
+    TwoComponentHamiltonian{T} <: AbstractHamiltonian{T}
+
+Abstract type for representing interacting two-component Hamiltonians in a Fock space with
+two different species. At least the following fields should be present:
+
+* `ha` Hamiltonian for species A
+* `hb` Hamiltonian for species B
+
+See [`AbstractHamiltonian`](@ref) for a list of methods that need to be defined.
+
+Provides and implementation of [`dimension`](@ref).
+"""
+abstract type TwoComponentHamiltonian{T} <: AbstractHamiltonian{T} end
+
+function dimension(::Type{T}, h::TwoComponentHamiltonian) where {T}
+    return dimension(T, h.ha) * dimension(T, h.hb)
+end
+
+"""
+    momentum(ham::AbstractHamiltonian)
+
+Momentum as a linear operator in Fock space. Pass a Hamiltonian `ham` in order to convey
+information about the Fock basis.
+
+Note: `momentum` is currently only defined on [`HubbardMom1D`](@ref).
+
+# Example
+
+```jldoctest
+julia> using Rimu.Hamiltonians: MomentumMom1D # MomentumMom1D is unexported
```

Comment:
This line does not make sense to me. Do we need to replace `MomentumMom1D` by `momentum`?

Reply:
Indeed, this docstring was written before I defined `momentum`. Removed the line because `momentum` is exported.